File: services/api/xtrace_api/schemas.py
"""Modelos del contrato REST de la API de búsqueda (PR-055/056 · FR-004/007/008/011
· UX-001 · contracts §1/§3/§4/§5).

`SearchResponse` reutiliza el JSON de la CLI `search` del spike (spec 001
contracts §1, FR-004): `search_id`, `processing_ms` y `results[]` con
`video_id`, `local_ref`, `match_score`, `matching_frames`,
`match_timestamp_ms` (nullable) y `evidence.visual`/`evidence.phash`.
La extensión **MAY** del contrato 003 §1 añade `title`/`page_url` (metadatos
de visualización, nullables) sin alterar los campos existentes.

`Stats` (contracts §3, FR-007) replica los campos de la CLI `stats` del
spike; `VideoCard` (contracts §4, FR-008) es la ficha del vídeo con sus
metadatos, fuente y enlace original.

`ApiError` es el cuerpo de error estructurado del contrato §5 (FR-011):
`{"error": "<mensaje en español>", "error_type": "<tipo-máquina>"}`.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class Evidence(BaseModel):
    """Evidencia de coincidencia de un resultado (paridad CLI, contracts §1)."""

    visual: float
    phash: float


class TimestampProvenance(BaseModel):
    """Procedencia trazable del timestamp mostrado (spec 006 · DATA-001)."""

    origin: Literal["base_index", "refined_asset"]
    status: Literal["improved", "unchanged", "unavailable", "limited", "disabled"]
    source: str | None = None
    asset_kind: Literal["thumbnail", "storyboard"] | None = None
    asset_url: str | None = None
    asset_position: int | None = None


class RefinementSummary(BaseModel):
    """Bounded aggregate metrics returned with a search response."""

    status: Literal["completed", "disabled", "unavailable", "limited", "failed"]
    candidates_requested: int = Field(ge=0)
    candidates_processed: int = Field(ge=0)
    assets_evaluated: int = Field(ge=0)
    assets_discarded: int = Field(ge=0)
    errors_count: int = Field(ge=0)
    bytes_downloaded: int = Field(ge=0)
    embedding_count: int = Field(ge=0)
    embedding_elapsed_ms: int = Field(ge=0)
    improved_results: int = Field(ge=0)
    elapsed_ms: int = Field(ge=0)

    @model_validator(mode="after")
    def validate_candidate_count(self) -> RefinementSummary:
        """Never expose a summary claiming more processed candidates than requested."""

        if self.candidates_processed > self.candidates_requested:
            raise ValueError("candidates_processed no puede superar candidates_requested")
        return self


class SearchResultItem(BaseModel):
    """Un vídeo rankeado de `POST /search` (contracts §1).

    `local_ref` puede ser `null` (backend in-memory, paridad CLI);
    `match_timestamp_ms` puede ser `null` (frame sin timestamp fiable,
    paridad FR-012 del spike); `title`/`page_url` pueden ser `null` (vídeos
    locales solo con `local_ref`; extensión MAY, FR-004).
    """

    video_id: str
    local_ref: str | None = None
    title: str | None = None
    page_url: str | None = None
    match_score: float
    matching_frames: int
    match_timestamp_ms: int | None = None
    evidence: Evidence
    timestamp_provenance: TimestampProvenance | None = None


class SearchResponse(BaseModel):
    """Respuesta 200 de `POST /search` (contracts §1, FR-004)."""

    search_id: str
    processing_ms: int = Field(ge=0)
    refinement: RefinementSummary | None = None
    results: list[SearchResultItem]


class Stats(BaseModel):
    """Métricas del índice (contracts §3 · FR-007): mismos campos que la CLI `stats`.

    `videos` = vídeos con ≥ 1 frame indexado (criterio de `VectorStore.stats()`
    del spike); `backend` es la etiqueta estable (`postgres` | `in-memory`) y
    `embedding_provider` el `model_id` del proveedor activo.
    """

    videos: int
    frames: int
    vectors: int
    backend: str
    embedding_provider: str


class VideoCard(BaseModel):
    """Ficha del vídeo (contracts §4 · FR-008): metadatos, fuente y enlace original.

    Nullables: `title`, `page_url`, `source` (vídeos locales sin fuente),
    `duration_ms`, `tags`, `published_at`, `thumbnail_url` (contracts §4).
    `source` es el nombre de la fuente (`sources.name`, join existente).
    """

    video_id: str
    local_ref: str
    title: str | None = None
    page_url: str | None = None
    source: str | None = None
    status: str
    duration_ms: int | None = None
    frame_count: int
    tags: list[str] | None = None
    published_at: datetime | None = None
    thumbnail_url: str | None = None
    excluded: bool


class ApiError(BaseModel):
    """Cuerpo de error estructurado (contracts §5 · FR-011 · UX-001).

    `error_type` es estable para consumo programático; `error` es el mensaje
    en español (idioma del frontend).
    """

    error: str
    error_type: str

File: src/lib/api/schemas.ts
import { z } from "zod";

/**
 * Contrato REST de `POST /search` (spec 003 · contracts/README.md §1 · paridad CLI FR-004).
 * El frontend valida la respuesta con zod: frontera estable entre frontend y API.
 *
 * Notas del contrato:
 * - `match_timestamp_ms` puede ser `null` (frame sin timestamp fiable).
 * - `local_ref` puede ser `null` (backend in-memory).
 * - `title`/`page_url` son la extensión MAY (metadatos de visualización): pueden faltar
 *   (`.nullish()`) o ser `null` — el frontend los trata como opcionales.
 */
/**
 * UUID con forma canónica (8-4-4-4-12 hexadecimal) sin exigir versión/variante:
 * el ejemplo del contrato §1 usa UUIDs de versión 0 (`...-0000-0000-...`) que zod v4
 * (`z.string().uuid()`) rechazaría; los ids reales de la API son uuid4 y también pasan.
 */
export const uuidShape = z
  .string()
  .regex(
    /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/,
    "debe ser un UUID",
  );

export const searchResultSchema = z.object({
  video_id: uuidShape,
  local_ref: z.string().nullable(),
  title: z.string().nullish(),
  page_url: z.string().url().nullish(),
  match_score: z.number(),
  matching_frames: z.number().int(),
  match_timestamp_ms: z.number().nullable(),
  evidence: z.object({
    visual: z.number(),
    phash: z.number(),
  }),
  timestamp_provenance: z
    .object({
      origin: z.enum(["base_index", "refined_asset"]),
      status: z.enum(["improved", "unchanged", "unavailable", "limited", "disabled"]),
      source: z.string().nullish(),
      asset_kind: z.enum(["thumbnail", "storyboard"]).nullish(),
      asset_url: z.string().url().nullish(),
      asset_position: z.number().int().nullish(),
    })
    .nullish(),
});

export const refinementSummarySchema = z
  .object({
    status: z.enum(["completed", "disabled", "unavailable", "limited", "failed"]),
    candidates_requested: z.number().int().nonnegative(),
    candidates_processed: z.number().int().nonnegative(),
    assets_evaluated: z.number().int().nonnegative(),
    assets_discarded: z.number().int().nonnegative(),
    errors_count: z.number().int().nonnegative(),
    bytes_downloaded: z.number().int().nonnegative(),
    embedding_count: z.number().int().nonnegative(),
    embedding_elapsed_ms: z.number().int().nonnegative(),
    improved_results: z.number().int().nonnegative(),
    elapsed_ms: z.number().int().nonnegative(),
  })
  .refine((summary) => summary.candidates_processed <= summary.candidates_requested, {
    message: "candidates_processed no puede superar candidates_requested",
    path: ["candidates_processed"],
  });

export const searchResponseSchema = z.object({
  search_id: uuidShape,
  processing_ms: z.number().int(),
  refinement: refinementSummarySchema.nullish(),
  results: z.array(searchResultSchema),
});

export type SearchResult = z.infer<typeof searchResultSchema>;
export type RefinementSummary = z.infer<typeof refinementSummarySchema>;
export type SearchResponse = z.infer<typeof searchResponseSchema>;
